RFDs/0001-job-worker-service.md

```diff
@@ -0,0 +1,468 @@
+}
+
+// ResourceManager manages system resources for multiple jobs.
+type ResourceManager struct {
+	mu   sync.Mutex              // Mutex for synchronizing access to the jobs map
+	jobs map[string]*jobResource // Map of job IDs to their associated resources
+}
+
+// NewResourceManager creates a new ResourceManager.
+func NewResourceManager() *ResourceManager
+
+// SetLimits sets resource limits for a job.
+func (rm *ResourceManager) SetLimits(jobID string, limits ResourceLimits) error
+
+// Adds a process to a job's cgroups.
+// It writes the process ID to the tasks file in each of the job's cgroups.
+func (rm *ResourceManager) AddProcess(jobID string, pid int) error
+
+// Retrieves the current resource usage for a job.
+func (rm *ResourceManager) GetUsage(jobID string) ResourceUsage
+
+// Removes a job's cgroups and cleans up associated resources.
+func (rm *ResourceManager) RemoveJob(jobID string) error
+```
+
+### 2. API
+
+#### gRPC API Specification
+
+```protobuf
+syntax = "proto3";
+
+package jobworker;
+
+import "google/protobuf/timestamp.proto";
+
+// JobWorker service provides methods to manage and monitor jobs.
+service JobWorker {
+  // Starts a new job with the given command and resource limits.
+  rpc StartJob(StartJobRequest) returns (StartJobResponse) {}
+  
+  // Stops a running job identified by its ID.
+  rpc StopJob(JobId) returns (StopJobResponse) {}
+  
+  // Retrieves the current status of a job.
+  rpc GetJobStatus(JobId) returns (stream JobStatusResponse) {}
+  
+  // Streams the output of a running job.
+  rpc GetJobOutput(JobId) returns (stream JobOutputResponse) {}
+
+  // Returns the list of all the jobs for the user
+  rpc ListJobs() returns (ListJobsResponse) {}
+}
+
+// StartJobRequest contains the information needed to start a new job.
+message StartJobRequest {
+  // The command to be executed.
+  string command = 1;
+  
+  // The arguments for the command.
+  repeated string args = 2;
+  
+  // The resource limits for the job.
+  ResourceLimits resource_limits = 3;
+}
+
+message JobId {
+  // The UUID of the job.
+  string id = 1;
+}
+
+// StartJobResponse contains the result of a job start request.
+message StartJobResponse {
+  // The UUID of the started job.
+  string id = 1;
+  
+  // The initial status of the job.
+  JobStatus status = 2;
+}
+
+// StopJobResponse contains the result of a job stop request.
+message StopJobResponse {
+  // The UUID of the stopped job.
+  string id = 1;
+  
+  // The final status of the job.
+  JobStatus status = 2;
+  
+  // The time when the job was started.
+  google.protobuf.Timestamp start_time = 3;
+  
+  // The time when the job was stopped.
+  google.protobuf.Timestamp stop_time = 4;
+}
+
+// JobStatusResponse contains the current status and resource usage of a job.
+message JobStatusResponse {
+  // The current status of the job.
+  JobStatus status = 1;
+  
+  // The current resource usage of the job.
+  ResourceUsage resource_usage = 2;
```

Comment:
Utilization reporting is not a challenge requirement

Reply:
I misunderstood. I had thought that the status of the job also meant the real time usage of the job. Removed.